File: Templates/Python/Regression/DecisionTreeRegression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 25 19:27:57 2019

@author: Kenil Shah
"""
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

dataset = pd.read_csv("Salary_Data.csv")

#Gives us a summary of the dataset
dataset.describe()

#to give column names
dataset.columns = []

#Diving the dataset into independent and dependent variables
X = dataset.iloc[:,:].values
Y = dataset.iloc[:,:].values

# Column names
list(dataset)

#Splitting the data into training set and validation set
from sklearn.cross_validation import train_test_split
train_X,test_X,train_Y,test_Y = train_test_split(X,Y,test_size = 0.2, random_state = 0) 

#Correlation Matrix
corr_matrix = train_X.corr()
f, ax = plt.subplots(figsize=(16,10))
ax = sns.heatmap(corr_matrix)
ax.set_title("correlation between all features")
figure = ax.get_figure()    

#Do feature scaling if required.
#Convert into categorical variable if required

#Model
from sklearn.tree import DecisionTreeRegressor
model = DecisionTreeRegressor(random_state = 0) 
# look for other features to tune the model according to your requirements

model.fit(train_X,train_Y)

#Predicting the values
y_pred = model.predict(test_X)

#If feature scaling is used on Y remember to inverse transform the y_pred variable

# Gives the coefficient values
model.coef_

#Get the intercept
model.intercept_

from sklearn.metrics import mean_squared_error,r2_score

#Mean Square Error
mse = mean_squared_error(test_Y,y_pred)

#Adjusted R Square
r2 = r2_score(test_Y,y_pred)

#Residual Plot to check how linear the model is
plt.scatter(model.predict(train_X),model.predict(train_X) - train_Y, c = 'b')
plt.hlines(y = 0,xmin = min(model.predict(train_X)),xmax = max(model.predict(train_X)))

#Visualize the Decision Tree
from sklearn.externals.six import StringIO  
from IPython.display import Image  
from sklearn.tree import export_graphviz
import pydotplus
dot_data = StringIO()
export_graphviz(model, out_file=dot_data,  
                filled=True, rounded=True,
                special_characters=True)
graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  
Image(graph.create_png())

#Saving the output file
y_pred = pd.Series(y_pred, name = 'target variable name')
submission = pd.DataFrame(data='The unique indentity of test dataset')
submission['Targer variable name'] = pred_y
submission.to_csv('submissionPath', sep='\t', index = False)

File: Templates/R/Regression/DecisionTreeRegression.R

dataset <- read.csv('Dataset Name')

summary(dataset) # provides a summary of the dataset

str(dataset) # gives us a structure of the dataset

names(dataset) # names of the column

library(caTools)

datasplit <- sample.split(dataset$DependentVariable,SplitRatio = 0.8)
train <- subset(dataset,datasplit = TRUE)
test <- sbuset(dataset,datasplit = FALSE)

#check for correlation 

#tocheck for two columns

cor(dataset&column1,dataset&column2,method = 'pearson') # other methods can be used

#between all the variables

library(ggpubr)
ggscatter(dataset,x= 'column 1',y = 'column 2',add = 'reg.line',
          conf.int = TRUE , cor.coef = TRUE , cor.method = 'pearson',
          xlab = 'Column 1',ylab = 'Column 2')

#Correltion diagram

library(corrplot)
x <- cor(dataset)
corrplot(x, type="upper", order="hclust")

#Correlation Matrix

corr.test(dataset)

#Correlation of one variable will every variable

library(corrr)
dataset %>% correlate() %>% focus(columnname)

#Do scaling if needed
#Convert into categorical variable if required

#Before creating the model, you might need to bin few values . You can use mapvalue function for that
library(plyr)
train$column <- mapvalues(train$column,from = c('A','B','C'),to = c('D','E','F')) # to map A to D, B to E,c to F
test$column <- mapvalues(test$column,from = c('A','B','C'),to = c('D','E','F')) # we need to map values in test as well

#Model
library(rpart)
model <- rpart(DependentVariable ~.,data = train,control = rpart.control(minsplit = 1))
                                                              # set various controls using rpart.control
?rpart.control # to check for all the control parameters

#Predicting values

predictedvalues <- predict(model,test)

#coefficients of the independent variables
model$coefficients

#Summary of the model will have the coefficients value, mean square error, adjusted r square and many other metrics

summary(model)

#Residual plot to check how linear the model is
plot(model$fitted.values,model$residuals)

#Visualing Decision Tree
library(rpart.plot)
rpart.plot(tree, box.palette="RdBu", shadow.col="gray", nn=TRUE)

#Saving the output file

write.csv(predictedvalues,'Filename',sep = '\t')
